__tests__/template/transaction/executor.test.ts

```diff
@@ -0,0 +1,617 @@
+  } else {
+    throw new Error('unexpected script');
+  }
+
+  expect(Object.keys(ctx.balance.balance)).toHaveLength(1);
+  expect(ctx.balance.balance[token]).toMatchObject({
+    tokens: -1n,
+    mint_authorities: 0,
+    melt_authorities: 0,
+  });
+};
+
+const TokenOutputExecutorTest = async executor => {
+  const ctx = new TxTemplateContext(getDefaultLogger(), DEBUG);
+  const interpreter = {
+    getNetwork: jest.fn().mockReturnValue(new Network('testnet')),
+  };
+  const ins = TokenOutputInstruction.parse({
+    type: 'output/token',
+    amount: 23,
+    address,
+    token,
+  });
+  await executor(interpreter, ctx, ins);
+
+  expect(interpreter.getNetwork).toHaveBeenCalledTimes(1);
+
+  expect(ctx.inputs).toHaveLength(0);
+  expect(ctx.tokens).toHaveLength(1);
+
+  expect(ctx.outputs).toHaveLength(1);
+  expect(ctx.outputs[0].value).toStrictEqual(23n);
+  expect(ctx.outputs[0].tokenData).toStrictEqual(1);
+  const script = ctx.outputs[0].parseScript(new Network('testnet'));
+  if (script && 'address' in script) {
+    expect(script.address.base58).toStrictEqual(address);
+  } else {
+    throw new Error('unexpected script');
+  }
+
+  expect(Object.keys(ctx.balance.balance)).toHaveLength(1);
+  expect(ctx.balance.balance[token]).toMatchObject({
+    tokens: -23n,
+    mint_authorities: 0,
+    melt_authorities: 0,
+  });
+};
+
+const AuthorityOutputExecutorTest = async executor => {
+  const ctx = new TxTemplateContext(getDefaultLogger(), DEBUG);
+  const interpreter = {
+    getNetwork: jest.fn().mockReturnValue(new Network('testnet')),
+  };
+  const ins = AuthorityOutputInstruction.parse({
+    type: 'output/authority',
+    authority: 'melt',
+    token,
+    address,
+  });
+  await executor(interpreter, ctx, ins);
+
+  expect(interpreter.getNetwork).toHaveBeenCalledTimes(1);
+
+  expect(ctx.inputs).toHaveLength(0);
+  expect(ctx.tokens).toHaveLength(1);
+
+  expect(ctx.outputs).toHaveLength(1);
+  expect(ctx.outputs[0].value).toStrictEqual(2n);
+  expect(ctx.outputs[0].tokenData).toStrictEqual(129);
+  const script = ctx.outputs[0].parseScript(new Network('testnet'));
+  if (script && 'address' in script) {
+    expect(script.address.base58).toStrictEqual(address);
+  } else {
+    throw new Error('unexpected script');
+  }
+
+  expect(Object.keys(ctx.balance.balance)).toHaveLength(1);
+  expect(ctx.balance.balance[token]).toMatchObject({
+    tokens: 0n,
+    mint_authorities: 0,
+    melt_authorities: -1,
+  });
+};
+
+const ShuffleExecutorTest = async executor => {
+  const ctx = new TxTemplateContext(getDefaultLogger(), DEBUG);
+  const interpreter = {};
+  const ins = ShuffleInstruction.parse({
+    type: 'action/shuffle',
+    target: 'all',
+  });
+  const arr: bigint[] = [];
+  for (let i = 1n; i < 10; i++) {
+    ctx.addOutput(-1, new Output(i, Buffer.alloc(1)));
+    ctx.addInput(-1, new Input(txId, Number(i)));
+    arr.push(i);
+  }
+  await executor(interpreter, ctx, ins);
```

Comment:
suggestion(non-blocking): Add assertions that ensure the steps are the same with a method like `contains`, or check the final balance of the transaction to confirm its equivalence to the original one.

Reply:
Done